File: sdks/python/abyss_sdk/event.py
"""Typed Agent event contract carried by plugin protocol version 1."""

from collections.abc import Mapping
from dataclasses import dataclass
from datetime import datetime
from typing import Any, Optional


@dataclass(frozen=True)
class DeviceContext:
    """Device context attached to an Agent event."""

    host_name: str
    platform: str
    os_version: Optional[str]


@dataclass(frozen=True)
class AgentContext:
    """Agent product context inferred from traffic."""

    name: str
    version: Optional[str]


@dataclass(frozen=True)
class LlmContext:
    """Provider and model context."""

    provider: str
    model: str


@dataclass(frozen=True)
class TokenUsage:
    """Normalized non-negative token counters."""

    input_tokens: int
    output_tokens: int
    cache_read_tokens: int
    cache_write_tokens: int
    reasoning_tokens: int
    total_tokens: int


@dataclass(frozen=True)
class ToolCall:
    """One normalized model-requested tool invocation."""

    call_id: str
    name: str
    input: str
    input_sha256: str


@dataclass(frozen=True)
class ToolResult:
    """One normalized tool result returned to the model."""

    call_id: str
    output: str
    output_sha256: str


@dataclass(frozen=True)
class ImageAttachment:
    """One policy-retained image attachment."""

    position: int
    media_type: str
    byte_size: int
    sha256: str
    content_base64: Optional[str]


@dataclass(frozen=True)
class AgentEvent:
    """One fully normalized broker Agent event."""

    event_id: str
    occurred_at: datetime
    device: DeviceContext
    agent: AgentContext
    session_id: str
    turn_index: int
    llm: LlmContext
    side: str
    text: Optional[str]
    token_usage: TokenUsage
    tool_calls: tuple[ToolCall, ...]
    tool_results: tuple[ToolResult, ...]
    attachments: tuple[ImageAttachment, ...]

    @classmethod
    def from_dict(cls, value: Mapping[str, Any]) -> "AgentEvent":
        """Decode and validate one protocol version 1 event."""

        event = _mapping(value, "AgentEvent")
        _reject_unknown(
            event,
            {
                "event_id",
                "occurred_at",
                "device",
                "agent",
                "session_id",
                "turn_index",
                "llm",
                "side",
                "text",
                "token_usage",
                "tool_calls",
                "tool_results",
                "attachments",
            },
            "AgentEvent",
        )
        side = _string(event, "side")
        if side not in {"request", "response"}:
            raise ValueError("AgentEvent.side must be request or response")
        turn_index = _integer(event, "turn_index")
        if turn_index < 1:
            raise ValueError("AgentEvent.turn_index must be greater than zero")
        occurred_at = _datetime(_string(event, "occurred_at"))
        return cls(
            event_id=_string(event, "event_id"),
            occurred_at=occurred_at,
            device=_device(_mapping(event.get("device"), "AgentEvent.device")),
            agent=_agent(_mapping(event.get("agent"), "AgentEvent.agent")),
            session_id=_string(event, "session_id"),
            turn_index=turn_index,
            llm=_llm(_mapping(event.get("llm"), "AgentEvent.llm")),
            side=side,
            text=_optional_string(event, "text"),
            token_usage=_token_usage(_mapping(event.get("token_usage"), "AgentEvent.token_usage")),
            tool_calls=tuple(
                _tool_call(item)
                for item in _list(event.get("tool_calls", []), "AgentEvent.tool_calls")
            ),
            tool_results=tuple(
                _tool_result(item)
                for item in _list(event.get("tool_results", []), "AgentEvent.tool_results")
            ),
            attachments=tuple(
                _attachment(item)
                for item in _list(event.get("attachments", []), "AgentEvent.attachments")
            ),
        )


def _device(value: Mapping[str, Any]) -> DeviceContext:
    _reject_unknown(value, {"host_name", "platform", "os_version"}, "AgentEvent.device")
    return DeviceContext(
        host_name=_string(value, "host_name"),
        platform=_string(value, "platform"),
        os_version=_optional_string(value, "os_version"),
    )


def _agent(value: Mapping[str, Any]) -> AgentContext:
    _reject_unknown(value, {"name", "version"}, "AgentEvent.agent")
    return AgentContext(
        name=_string(value, "name"),
        version=_optional_string(value, "version"),
    )


def _llm(value: Mapping[str, Any]) -> LlmContext:
    _reject_unknown(value, {"provider", "model"}, "AgentEvent.llm")
    return LlmContext(provider=_string(value, "provider"), model=_string(value, "model"))


def _token_usage(value: Mapping[str, Any]) -> TokenUsage:
    fields = {
        "input_tokens",
        "output_tokens",
        "cache_read_tokens",
        "cache_write_tokens",
        "reasoning_tokens",
        "total_tokens",
    }
    _reject_unknown(value, fields, "AgentEvent.token_usage")
    counters: dict[str, int] = {}
    for field in fields:
        counter = _integer(value, field)
        if counter < 0:
            raise ValueError(f"AgentEvent.token_usage.{field} must not be negative")
        counters[field] = counter
    return TokenUsage(**counters)


def _tool_call(value: Any) -> ToolCall:
    item = _mapping(value, "AgentEvent.tool_calls[]")
    _reject_unknown(item, {"call_id", "name", "input", "input_sha256"}, "tool call")
    return ToolCall(
        call_id=_string(item, "call_id"),
        name=_string(item, "name"),
        input=_string(item, "input"),
        input_sha256=_string(item, "input_sha256"),
    )


def _tool_result(value: Any) -> ToolResult:
    item = _mapping(value, "AgentEvent.tool_results[]")
    _reject_unknown(item, {"call_id", "output", "output_sha256"}, "tool result")
    return ToolResult(
        call_id=_string(item, "call_id"),
        output=_string(item, "output"),
        output_sha256=_string(item, "output_sha256"),
    )


def _attachment(value: Any) -> ImageAttachment:
    item = _mapping(value, "AgentEvent.attachments[]")
    _reject_unknown(
        item,
        {"position", "media_type", "byte_size", "sha256", "content_base64"},
        "attachment",
    )
    media_type = _string(item, "media_type")
    if media_type not in {"image/png", "image/jpeg", "image/webp", "image/gif"}:
        raise ValueError(f"unsupported AgentEvent attachment media type {media_type}")
    position = _integer(item, "position")
    byte_size = _integer(item, "byte_size")
    if position < 0 or byte_size < 0:
        raise ValueError("AgentEvent attachment sizes and positions must not be negative")
    return ImageAttachment(
        position=position,
        media_type=media_type,
        byte_size=byte_size,
        sha256=_string(item, "sha256"),
        content_base64=_optional_string(item, "content_base64"),
    )


def _mapping(value: Any, label: str) -> Mapping[str, Any]:
    if not isinstance(value, Mapping):
        raise TypeError(f"{label} must be an object")
    return value


def _list(value: Any, label: str) -> list[Any]:
    if not isinstance(value, list):
        raise TypeError(f"{label} must be an array")
    return value


def _string(value: Mapping[str, Any], field: str) -> str:
    result = value.get(field)
    if not isinstance(result, str):
        raise TypeError(f"{field} must be a string")
    return result


def _optional_string(value: Mapping[str, Any], field: str) -> Optional[str]:
    result = value.get(field)
    if result is None:
        return None
    if not isinstance(result, str):
        raise TypeError(f"{field} must be a string when present")
    return result


def _integer(value: Mapping[str, Any], field: str) -> int:
    result = value.get(field)
    if isinstance(result, bool) or not isinstance(result, int):
        raise TypeError(f"{field} must be an integer")
    return result


def _datetime(value: str) -> datetime:
    try:
        return datetime.fromisoformat(value.replace("Z", "+00:00"))
    except ValueError as error:
        raise ValueError("AgentEvent.occurred_at must be an RFC 3339 timestamp") from error


def _reject_unknown(value: Mapping[str, Any], allowed: set, label: str) -> None:
    unknown = set(value) - allowed
    if unknown:
        raise ValueError(f"{label} contains unknown fields: {sorted(unknown)}")

File: crates/abyss-plugin-protocol/src/event.rs
//! Language-neutral Agent event wire types shared by the broker and Rust SDK.
//!
//! These types are independent of the current backend ingest request. A
//! delivery plugin may translate an event into its destination's API without
//! making that remote API part of the broker contract.
//! Agent event compatibility is part of the broker plugin protocol version, so
//! an individual event does not carry a separate schema version.

use std::num::NonZeroU32;

use chrono::{DateTime, Utc};
use serde::{Deserialize, Deserializer, Serialize, Serializer};
/// One event decoded with the contract defined by the active plugin protocol.
#[derive(Debug, Deserialize, Serialize)]
#[serde(deny_unknown_fields)]
pub struct AgentEvent {
    /// Stable event identifier assigned by the event producer.
    pub event_id: String,
    /// Time at which the broker observed the logical event.
    pub occurred_at: DateTime<Utc>,
    /// Device context observed by the broker.
    pub device: DeviceContext,
    /// Agent product context inferred from traffic.
    pub agent: AgentContext,
    /// Provider or collector session identifier.
    pub session_id: String,
    /// One-based collector turn index within the session.
    pub turn_index: NonZeroU32,
    /// Provider and model context.
    pub llm: LlmContext,
    /// Request or response side represented by this event.
    pub side: AgentEventSide,
    /// Policy-retained request or response text.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub text: Option<String>,
    /// Token counters attributed to this event side.
    pub token_usage: TokenUsage,
    /// Normalized tool calls associated with this event side.
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub tool_calls: Vec<ToolCall>,
    /// Normalized tool results associated with this event side.
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub tool_results: Vec<ToolResult>,
    /// Policy-retained image attachments.
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub attachments: Vec<ImageAttachment>,
}

/// Device context attached to one Agent event.
#[derive(Debug, Deserialize, Serialize)]
#[serde(deny_unknown_fields)]
pub struct DeviceContext {
    /// Human-readable host name.
    pub host_name: String,
    /// Platform namespace such as `linux`, `macos`, or `windows`.
    pub platform: String,
    /// Operating-system version when available.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub os_version: Option<String>,
}

/// Agent product context attached to one Agent event.
#[derive(Debug, Deserialize, Serialize)]
#[serde(deny_unknown_fields)]
pub struct AgentContext {
    /// Product name such as `codex` or `claude-code`.
    pub name: String,
    /// Agent version when it can be inferred from traffic.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub version: Option<String>,
}

/// LLM provider and model context attached to one Agent event.
#[derive(Debug, Deserialize, Serialize)]
#[serde(deny_unknown_fields)]
pub struct LlmContext {
    /// Provider namespace with typed variants for built-in providers.
    pub provider: LlmProvider,
    /// Provider model name.
    pub model: String,
}

/// Known and extension LLM provider namespaces.
#[derive(Debug, Eq, PartialEq)]
#[non_exhaustive]
pub enum LlmProvider {
    /// OpenAI-compatible first-party API traffic.
    OpenAi,
    /// Anthropic first-party API traffic.
    Anthropic,
    /// Provider namespace not yet represented by a built-in variant.
    Other(String),
}

impl LlmProvider {
    /// Creates a provider from its wire namespace.
    #[must_use]
    pub fn from_wire_name(provider: String) -> Self {
        match provider.as_str() {
            "openai" => Self::OpenAi,
            "anthropic" => Self::Anthropic,
            _ => Self::Other(provider),
        }
    }

    /// Returns the provider namespace carried on the wire.
    #[must_use]
    pub fn wire_name(&self) -> &str {
        match self {
            Self::OpenAi => "openai",
            Self::Anthropic => "anthropic",
            Self::Other(provider) => provider,
        }
    }
}

impl Serialize for LlmProvider {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        serializer.serialize_str(self.wire_name())
    }
}

impl<'de> Deserialize<'de> for LlmProvider {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        String::deserialize(deserializer).map(Self::from_wire_name)
    }
}

/// Request or response side represented by one usage event.
#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[non_exhaustive]
#[serde(rename_all = "snake_case")]
pub enum AgentEventSide {
    /// Agent request content and input-side token usage.
    Request,
    /// Provider response content and output-side token usage.
    Response,
}

impl AgentEventSide {
    /// Returns the request/response name carried by backend translations.
    #[must_use]
    pub const fn wire_name(self) -> &'static str {
        match self {
            Self::Request => "request",
            Self::Response => "response",
        }
    }
}

/// One normalized model-requested tool invocation.
#[derive(Debug, Deserialize, Serialize)]
#[serde(deny_unknown_fields)]
pub struct ToolCall {
    /// Broker-normalized identifier shared with the corresponding tool result.
    pub call_id: String,
    /// Broker-normalized tool name, such as `Bash`, `Read`, or `exec`.
    pub name: String,
    /// Complete provider-visible tool input normalized as text by the broker.
    pub input: String,
    /// Lowercase SHA-256 digest of `input`.
    pub input_sha256: String,
}

/// One normalized result submitted back to the model after a tool call.
#[derive(Debug, Deserialize, Serialize)]
#[serde(deny_unknown_fields)]
pub struct ToolResult {
    /// Broker-normalized identifier shared with the originating tool call.
    pub call_id: String,
    /// Provider-visible result normalized as text by the broker.
    pub output: String,
    /// Lowercase SHA-256 digest of `output`.
    pub output_sha256: String,
}

/// Normalized non-negative token counters.
#[derive(Debug, Deserialize, Serialize)]
#[serde(deny_unknown_fields)]
pub struct TokenUsage {
    /// Tokens consumed by prompt or input content.
    pub input_tokens: u64,
    /// Tokens produced by model output.
    pub output_tokens: u64,
    /// Input tokens read from a provider cache.
    pub cache_read_tokens: u64,
    /// Input tokens written into a provider cache.
    pub cache_write_tokens: u64,
    /// Tokens spent in model reasoning traces.
    pub reasoning_tokens: u64,
    /// Provider total or a total derived by the producer.
    pub total_tokens: u64,
}

/// One policy-retained image attachment.
#[derive(Debug, Deserialize, Serialize)]
#[serde(deny_unknown_fields)]
pub struct ImageAttachment {
    /// Stable zero-based presentation position within the event.
    pub position: u32,
    /// Validated image media type.
    pub media_type: ImageMediaType,
    /// Decoded byte count before base64 transport encoding.
    pub byte_size: u64,
    /// Lowercase SHA-256 digest of the decoded image bytes.
    pub sha256: String,
    /// Image content when policy permits full image retention.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub content_base64: Option<String>,
}

/// Image media types supported by broker plugin protocol version 1.
#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[non_exhaustive]
pub enum ImageMediaType {
    /// Portable Network Graphics.
    #[serde(rename = "image/png")]
    Png,
    /// Joint Photographic Experts Group image.
    #[serde(rename = "image/jpeg")]
    Jpeg,
    /// WebP image.
    #[serde(rename = "image/webp")]
    Webp,
    /// Graphics Interchange Format image.
    #[serde(rename = "image/gif")]
    Gif,
}

impl ImageMediaType {
    /// Returns the MIME type carried on the wire.
    #[must_use]
    pub const fn wire_name(self) -> &'static str {
        match self {
            Self::Png => "image/png",
            Self::Jpeg => "image/jpeg",
            Self::Webp => "image/webp",
            Self::Gif => "image/gif",
        }
    }
}

#[cfg(test)]
mod tests {
    use std::num::NonZeroU32;

    use super::{
        AgentContext, AgentEvent, AgentEventSide, DeviceContext, LlmContext, LlmProvider,
        TokenUsage, ToolCall, ToolResult,
    };
    use chrono::{TimeZone as _, Utc};

    #[test]
    fn agent_event_serializes_as_one_flat_typed_payload() {
        let event = sample_event(LlmProvider::OpenAi);
        let serialized = serde_json::to_value(event).expect("Agent event should serialize");

        assert!(
            serialized.get("schema_version").is_none(),
            "Agent events should be versioned by the plugin protocol"
        );
        assert_eq!(
            serialized["side"], "request",
            "the event side should be a direct typed field"
        );
        assert!(
            serialized.get("event_type").is_none() && serialized.get("payload").is_none(),
            "a single event kind should not add a speculative discriminator or payload wrapper"
        );
        assert!(
            serialized.get("metadata").is_none(),
            "the public event must not expose an unstructured metadata object"
        );
        assert!(
            serialized.get("text").is_none(),
            "absent optional content should be omitted"
        );
    }

    #[test]
    fn tool_activity_serializes_as_named_structures() {
        let mut event = sample_event(LlmProvider::OpenAi);
        event.tool_calls.push(ToolCall {
            call_id: "call-1".to_owned(),
            name: "exec".to_owned(),
            input: "pwd".to_owned(),
            input_sha256: "call-hash".to_owned(),
        });
        event.tool_results.push(ToolResult {
            call_id: "call-1".to_owned(),
            output: "/workspace".to_owned(),
            output_sha256: "result-hash".to_owned(),
        });

        let serialized = serde_json::to_value(event).expect("Agent event should serialize");

        assert_eq!(serialized["tool_calls"][0]["name"], "exec");
        assert_eq!(serialized["tool_calls"][0]["input"], "pwd");
        assert_eq!(serialized["tool_results"][0]["call_id"], "call-1");
        assert_eq!(serialized["tool_results"][0]["output"], "/workspace");
    }

    #[test]
    fn agent_event_rejects_an_unstructured_metadata_object() {
        let mut serialized = serde_json::to_value(sample_event(LlmProvider::OpenAi))
            .expect("event should serialize");
        serialized["metadata"] = serde_json::json!({"provider_private_field": true});

        let error = serde_json::from_value::<AgentEvent>(serialized)
            .expect_err("undeclared metadata should be rejected");

        assert!(
            error.to_string().contains("unknown field `metadata`"),
            "error should identify metadata as outside the public contract"
        );
    }

    #[test]
    fn extension_provider_namespace_round_trips() {
        let serialized = serde_json::to_value(sample_event(LlmProvider::Other(
            "customer-private-llm".to_owned(),
        )))
        .expect("Agent event should serialize");
        let decoded: AgentEvent =
            serde_json::from_value(serialized).expect("Agent event should deserialize");

        assert_eq!(
            decoded.llm.provider,
            LlmProvider::Other("customer-private-llm".to_owned()),
            "unknown provider namespaces should be preserved"
        );
    }

    fn sample_event(provider: LlmProvider) -> AgentEvent {
        AgentEvent {
            event_id: "evt-test".to_owned(),
            occurred_at: Utc
                .with_ymd_and_hms(2026, 8, 19, 10, 0, 0)
                .single()
                .expect("sample timestamp should be valid"),
            device: DeviceContext {
                host_name: "test-host".to_owned(),
                platform: "macos".to_owned(),
                os_version: None,
            },
            agent: AgentContext {
                name: "codex".to_owned(),
                version: None,
            },
            session_id: "session-1".to_owned(),
            turn_index: NonZeroU32::new(1).expect("one should be non-zero"),
            llm: LlmContext {
                provider,
                model: "gpt-test".to_owned(),
            },
            side: AgentEventSide::Request,
            text: None,
            token_usage: TokenUsage {
                input_tokens: 10,
                output_tokens: 0,
                cache_read_tokens: 0,
                cache_write_tokens: 0,
                reasoning_tokens: 0,
                total_tokens: 10,
            },
            tool_calls: Vec::new(),
            tool_results: Vec::new(),
            attachments: Vec::new(),
        }
    }
}
